ffs-shared/rule/src/commonMain/kotlin/rule/RuleEval.kt

```diff
@@ -571,19 +547,142 @@ private sealed class RuleExpr<out T> {
+                }
+                override fun iterator(): Iterator<T> {
+                    TODO("Doesn't need this function")
+                }
+            }
+        }
+    }
+}
+
+internal class Ip {
+    internal data class Numeric(val ipString: String) : Comparable<Numeric> {
+        val value: BigInteger by lazy {
+            val segments = ipString.expandIpv6().toSegments()
+
+            if (segments.size > 4) {
+                segments.fold(BigInteger.fromInt(0)) { acc, seg -> (acc shl 16) + seg.toInt() }
+            } else {
+                segments.fold(BigInteger.fromInt(0)) { acc, octet -> (acc shl 8) + octet.toShort() }
+            }
+        }
+
+        override fun compareTo(other: Numeric) = value.compareTo(other.value)
+    }
+
+    internal data class IpRange(val cidr: String) : ClosedRange<Numeric> {
+        private lateinit var min: Numeric
+        private lateinit var max: Numeric
+
+        init {
+            val ipWidth = cidr.split('/')
+            val segments = ipWidth[0].expandIpv6().toSegments()
+
+            if (segments.size > 4) {
+                val width = ipWidth.getOrNull(1)?.toInt() ?: 128
+                calculateIpv6Range(segments, width)
+            } else {
+                val width = ipWidth.getOrNull(1)?.toByte() ?: 32
+                calculateIpv4Range(segments, width)
+            }
+        }
+
+        private fun calculateIpv6Range(segments: List<UShort>, width: Int) {
+            val min = arrayOf<UShort>(0u, 0u, 0u, 0u, 0u, 0u, 0u, 0u)
+            val max = arrayOf<UShort>(0u, 0u, 0u, 0u, 0u, 0u, 0u, 0u)
+
+            val fullSegments = width / 16
+            val remainder = width % 16
+            segments.forEachIndexed { i, value ->
+                val currentSubnetMask = if (i < fullSegments) {
+                    (0xFFFF).toUShort()
+                } else if (i == fullSegments) {
+                    (0xFFFF shl (16 - remainder)).toUShort()
+                } else {
+                    0u
+                }
+
+                min[i] = value and currentSubnetMask
+                max[i] = value or currentSubnetMask.inv()
+            }
+
+            this.min = Numeric(min.joinToString(":") { it.toString(16) })
+            this.max = Numeric(max.joinToString(":") { it.toString(16) })
+        }
+
+        private fun calculateIpv4Range(octets: List<UShort>, width: Byte) {
+            val mask = 0xFFFFFFFF shl (32 - width)
+            val subnet =
+                arrayOf(mask shr 24, mask shr 16, mask shr 8, mask).map { it.toUShort() }
+
+            val min = arrayOf<UByte>(0u, 0u, 0u, 0u)
+            val max = arrayOf<UByte>(0u, 0u, 0u, 0u)
+            octets.forEachIndexed { i, value ->
+                min[i] = (value and subnet[i]).toUByte()
+                max[i] = (value or subnet[i].inv()).toUByte()
             }
+
+            this.min = Numeric(min.joinToString("."))
+            this.max = Numeric(max.joinToString("."))
+        }
```

Comment:
Just a gut feeling, but did you consider having a single `calculateIpRange`? We can infer the length from the `octets` size, generalizing the calculations for any size, and possibly reuse the same logic. I think the only unknown is the base size, which is 10 for ipv4 and 16 for ipv6. That could be a parameter, to format the output.

Reply:
>I assume we're using UShort to represent IPv6's quartets. Did you consider using UByte and representing IPv6 using its octets instead, so that the underlying logic is the same across both versions?

Related to this⬆️ comment, if we generalize the IP to octets, Yes, we can use one function to do both types because the mask calculation logic is exactly the same